
# --- UI Layout with fixedPage() ---
ui <- fixedPage(theme = shinytheme("sandstone"),
                
    # --- Global CSS Styling ---
    tags$style(HTML("
    .container {
      max-width: 1600px;
      margin-left: auto;
      margin-right: auto;
    }
    #map {
      width: 100% !important;
      height: 50vh;  /* Set the height to 50% of the viewport height */
    }
    .centered-box {
      display: flex;
      justify-content: center;
      align-items: center;
    }
    .center-content {
      margin: auto;
      text-align: center;
    }
    .checkbox-inline {
      margin-bottom: 5px;
    }
    #contour_checkbox,
    #highlight_high_sd {
      margin-right: 10px;
    }
    
    /* Title Panel color */
    .title-panel {
      background-color: #2e2d2b;  /* color */
      color: white;               /* White text color */
      padding: 10px;
      text-align: center;
      font-size: 24px;
      font-weight: bold;
    }
    
    /* Button Highlight in color */
    .btn-color {
      background-color: #749c41;  /* color */
      color: white;               /* White text color */
      border: none;
      font-weight: bold;
    }
    .btn-color:hover {
      background-color: #9c4141;  /* Darker color on hover */
      color: white;
    }
  ")),  # End of Global CSS Styling
                
   # --- Application Title ---
   div(class = "title-panel", "Display the results of your CEPHALOPOD run !"),  # End of Application Title
   
   # --- Sidebar and Main Panel Layout ---
   sidebarLayout(
     
     # --- Sidebar Panel ---
     sidebarPanel(
       width = 4,  # Sidebar takes up 4 out of 12 grid columns (300px with fixedPage)
       
       radioButtons("data_type", "1. Select Data Type", choices = c("Diversity", "Ensemble")),
       p("Select", tags$b("Diversity"), "to display diversity estimations.", tags$br(), "Select", tags$b("Ensemble"), "to display taxa-level ensemble projections."),
       tags$hr(style="border-color: black;"),
       
       fileInput("ncfile", "2. Upload the appropriate NetCDF File", accept = ".nc"),
       p("Please select the file as generated by CEPHALOPOD (e.g.", tags$b("output_diversity.nc"), ")."),
       tags$hr(style="border-color: black;"),
       
       uiOutput("feature_select"),  # Dynamic UI for feature selection
       p("Please select the variable to display. This corresponds to the diversity estimates or species names present in the .nc file."),
       tags$hr(style="border-color: black;"),
       
       selectInput("variable", "4. Select Variable", choices = c("mean_values", "sd_values")),
       p("Both the yearly average and standard deviation across month are available for display. Select the one you are interested in."),
       tags$hr(style="border-color: black;"),
       
       sliderInput("time", "5. Select Time", min = 1, max = 13, value = 1),
       p("Select the month to display with Jan (1), Dec(12), and Annual Average (13)", tags$br(), textOutput("time_label")),
       tags$hr(style="border-color: black;"),
       
       checkboxInput("contour_checkbox", "Show Contours", value = FALSE),
       p("Select to show the isolines on the map"),
       
       checkboxInput("highlight_high_sd", "Highlight High SD Values", value = FALSE),
       p("Select to highlight pixels presenting a standard deviation above 0.5"),
       
       uiOutput("highlight_type_ui"),
       tags$hr(style="border-color: black;"),
       
       actionButton("update", "Update Map", class = "btn-color")  # Add class for color button
     ),  # End of Sidebar Panel
     
     # --- Main Panel ---
     mainPanel(
       width = 8,  # Main panel takes up 8 out of 12 grid columns (700px with fixedPage)
       tabsetPanel(
         id = "tabselected",
         
         # --- Explanations Tab ---
         tabPanel(
           "General informations",
           fluidRow(
               hr(),
               p(tags$b("CEPHALOPOD"), "is a ensemble habitat modelling pipeline designed to produce standardized, quality checked and inter-comparable outputs across different data types, including occurrences, abundances, biomass and metagenomic reads.",
                 tags$br(),
                 "A detailed description of the CEPHALOPOD pipeline is available in this ", a(href="https://github.com/alexschickele/CEPHALOPOD/tree/main", "github repository"),
                 tags$br(),
                 "The present R Shiny application is designed as a user-friendly visualization tool for the", tags$b("output_diversity.nc"),  "and", tags$b("output_ensemble.nc"),  
                 "produced by the pipeline. Follow the step-wise instructions in the left-panel to do so"),
               tags$br(),
               tags$hr(style="border-color: black;"),
              
           )  # End of fluidRow for explanations
         ),  # End of Explanations Tab
         
         # --- Map Tab ---
         tabPanel(
           "Map",
           fluidRow(
               h3("Spatial projection"),
               leafletOutput("map", height = "325px")
           ),  # End of fluidRow for map display
           fluidRow(
               plotOutput("legend_plot", height = 100)
             ),  # End of fluidRow for legend plot
           fluidRow(
               h3("Additional information"),
               textOutput("explanation_text"),
               uiOutput("ensemble_box")
           )  # End of fluidRow for additional map-related content
         )  # End of Map Tab
         
       )  # End of tabsetPanel
     )  # End of Main Panel
     
   )  # End of Sidebar and Main Panel Layout
   
)  # End of UI Layout with fixedPage()
